minio/api.py

```diff
@@ -176,6 +182,21 @@ def __init__(self, endpoint, access_key=None,
         self._user_agent = _DEFAULT_USER_AGENT
         self._trace_output_stream = None
 
+        # Set credentials if possible
+
+        if credentials is not None:
+            self._credentials = credentials
+        else:
+            self._credentials = Credentials(
+                provider= Chain(
+                    providers=[
+                        Static(access_key, secret_key, session_token),
+                        EnvAWS(),
+                        EnvMinio(),
```

Comment:
Can I request a small change of adding `IamEc2MetaData` to the end of the provider chain? I can't see it causing any issues.
This will allow container images that already use minio-py to pick up the AWS credentials without the user having to modify the image of pass api keys (granted the maintainer will have to update the library). 
(In my use case it is a 3rd party maintained container deployed to Kubernetes on AWS)

Reply:
I understand! When I add the `IamEc2MetaData` credentials to the chain provider most of the tests brake. I get the following error:
```
======================================================================
ERROR: test_stat_object_works (tests.unit.stat_object_test.StatObject)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/febg/Documents/Work/CodeCov/minio-py/venv/lib/python3.7/site-packages/mock/mock.py", line 1330, in patched
    return func(*args, **keywargs)
  File "/Users/febg/Documents/Work/CodeCov/minio-py/tests/unit/stat_object_test.py", line 58, in test_stat_object_works
    client.stat_object('hello', 'world')
  File "/Users/febg/Documents/Work/CodeCov/minio-py/minio/api.py", line 1067, in stat_object
    object_name=object_name, headers=headers)
  File "/Users/febg/Documents/Work/CodeCov/minio-py/minio/api.py", line 1922, in _url_open
    region = self._get_bucket_region(bucket_name)
  File "/Users/febg/Documents/Work/CodeCov/minio-py/minio/api.py", line 1852, in _get_bucket_region
    region = self._get_bucket_location(bucket_name)
  File "/Users/febg/Documents/Work/CodeCov/minio-py/minio/api.py", line 1876, in _get_bucket_location
    if self._credentials.get().access_key is None or self._credentials.get().secret_key is None:
  File "/Users/febg/Documents/Work/CodeCov/minio-py/minio/credentials/credentials.py", line 61, in get
    creds = self._provider.retrieve()
  File "/Users/febg/Documents/Work/CodeCov/minio-py/minio/credentials/chain.py", line 27, in retrieve
    creds = provider.retrieve()
  File "/Users/febg/Documents/Work/CodeCov/minio-py/minio/credentials/aws_iam.py", line 62, in retrieve
    creds_list = self.request_cred_list()
  File "/Users/febg/Documents/Work/CodeCov/minio-py/minio/credentials/aws_iam.py", line 43, in request_cred_list
    res = self._http_client.urlopen('GET', url)
TypeError: urlopen() missing 1 required positional argument: 'headers'
```
This is weird since 'headers' is not required, and I do not get this error when I test the `IamEc2MetaData` provider by itself!

I went ahead and added the headers parameter (`headers=none`) and then I get this error:
```
======================================================================
FAIL: test_stat_object_works (tests.unit.stat_object_test.StatObject)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/febg/Documents/Work/CodeCov/minio-py/venv/lib/python3.7/site-packages/mock/mock.py", line 1330, in patched
    return func(*args, **keywargs)
  File "/Users/febg/Documents/Work/CodeCov/minio-py/tests/unit/stat_object_test.py", line 58, in test_stat_object_works
    client.stat_object('hello', 'world')
  File "/Users/febg/Documents/Work/CodeCov/minio-py/minio/api.py", line 1067, in stat_object
    object_name=object_name, headers=headers)
  File "/Users/febg/Documents/Work/CodeCov/minio-py/minio/api.py", line 1922, in _url_open
    region = self._get_bucket_region(bucket_name)
  File "/Users/febg/Documents/Work/CodeCov/minio-py/minio/api.py", line 1852, in _get_bucket_region
    region = self._get_bucket_location(bucket_name)
  File "/Users/febg/Documents/Work/CodeCov/minio-py/minio/api.py", line 1876, in _get_bucket_location
    if self._credentials.get().access_key is None or self._credentials.get().secret_key is None:
  File "/Users/febg/Documents/Work/CodeCov/minio-py/minio/credentials/credentials.py", line 61, in get
    creds = self._provider.retrieve()
  File "/Users/febg/Documents/Work/CodeCov/minio-py/minio/credentials/chain.py", line 27, in retrieve
    creds = provider.retrieve()
  File "/Users/febg/Documents/Work/CodeCov/minio-py/minio/credentials/aws_iam.py", line 62, in retrieve
    creds_list = self.request_cred_list()
  File "/Users/febg/Documents/Work/CodeCov/minio-py/minio/credentials/aws_iam.py", line 43, in request_cred_list
    res = self._http_client.urlopen('GET', url, headers=None)
  File "/Users/febg/Documents/Work/CodeCov/minio-py/tests/unit/minio_mocks.py", line 77, in urlopen
    return self.request(method, url, headers)
  File "/Users/febg/Documents/Work/CodeCov/minio-py/tests/unit/minio_mocks.py", line 71, in request
    return_request.mock_verify(method, url, headers)
  File "/Users/febg/Documents/Work/CodeCov/minio-py/tests/unit/minio_mocks.py", line 46, in mock_verify
    eq_(self.method, method)
AssertionError: 'HEAD' != 'GET'
```

Which makes me believe we need to mock the response from AWS metadata service!

Any thoughts on how to fix this?

Reply:
looks like the headers cannot be none @febg. Can you verify if you are using IMDSv2 version - If so, you need to set token headers "X-aws-ec2-metadata-token: $TOKEN"

Reply:
Is there any API reference link? Can you paste it here?

Reply:
I believe this is the reference doc for AWS metadata service
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html

Reply:
We are using IMDSv1 version. It works fine when I test the credentials provider inside an EC2 instance. We could just mock the response, but we would need to mock for 27 tests

Reply:
Then shouldn't the `iam_security_creds_path = '/latest/meta-data/iam/security-credentials'` be `iam_security_creds_path = '/latest/meta-data/iam/security-credentials'/{role-name}`

Reply:
Did you try using `'HEAD'` for listing role-names - `request_cred_list()`

Reply:
Mocking the responses will be a burden, You could remove `IamEc2MetaData` from the list of provides in that case. IMO.

Reply:
I will go for removing `IamEc2MetaData`. I havent been able to fix this error

Reply:
I found a bug in the IamEc2MetaData implementation - it threw an exception when it couldn't access the metadata service instead of returning no credentials. This was causing the tests to fail that were doing anonymous access calls to get the bucket location (and others).
Here's a patch to fix it https://gist.github.com/devgrok/6dd9afce7ad6eee31aa9dd82d2a196c6

An additional change in the patch: I added setuptools to automated builds because in one virtualenv, configparser failed to install without it

Reply:
@devgrok I tried to apply the patch but it seems it is corrupted! Ill add the changes manually